public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx

```diff
@@ -62,32 +59,15 @@ export class VizPanelManager extends SceneObjectBase<VizPanelManagerState> {
 
   private _dataObjectSubscription: Unsubscribable | undefined;
 
-  public constructor(panel: VizPanel, dashboardRef: SceneObjectRef<DashboardScene>) {
+  public constructor(panel: VizPanel) {
     super({ panel });
 
-    /**
-     * If the panel uses a shared query, we clone the source runner and attach it as a data provider for the shared one.
-     * This way the source panel does not to be present in the edit scene hierarchy.
-     */
-    if (panel.state.$data instanceof ShareQueryDataProvider) {
-      const sharedProvider = panel.state.$data;
-      if (sharedProvider.state.query.panelId) {
-        const keyToFind = getVizPanelKeyForPanelId(sharedProvider.state.query.panelId);
-        const source = findObjectInScene(dashboardRef.resolve(), (scene: SceneObject) => scene.state.key === keyToFind);
-        if (source) {
-          sharedProvider.setState({
-            $data: source.state.$data!.clone(),
-          });
-        }
-      }
-    }
-
+    const dataProvider = this.state.panel.state.$data;
+    dataProvider?.addActivationHandler(() => this.setupDataObjectSubscription());
```

Comment:
one thing I don't understand is  this.setupDataObjectSubscription(), It subscribes query runner state to detect datasource changes, and calls loadDataSource when one is detected. But all data source changes go through changePanelDataSource and it calls this.loadDataSource(). So what scenario is setupDataObjectSubscription for?

Reply:
It's for a scenario when we don't replace data provider object, but update it's state. Think for instance changing from one prometheus ds to another. But now when I think about this we can handle it imperatively in the data source change handler by just calling `loadDataSource` in this scenario, which we don't do currently.